### PR TITLE
Account for possible concurrent topic creation in PubSub topic auto-c…

### PR DESCRIPTION
**Pull request by mvmn**
…reate functionality.

https://github.com/spring-cloud/spring-cloud-gcp/issues/2410

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@mvmn Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-gcp&pullRequestId=2411)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-gcp&pullRequestId=2411) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

File: spring-cloud-gcp-pubsub-stream-binder/src/main/java/org/springframework/cloud/gcp/stream/binder/pubsub/provisioning/PubSubChannelProvisioner.java [hunk 1]
```diff
@@ -118,7 +120,11 @@ private Topic makeSureTopicExists(String topicName, boolean autoCreate) {
 		Topic topic = this.pubSubAdmin.getTopic(topicName);
 		if (topic == null) {
 			if (autoCreate) {
-				topic = this.pubSubAdmin.createTopic(topicName);
+				try {
+					topic = this.pubSubAdmin.createTopic(topicName);
+				} catch(AlreadyExistsException alreadyExistsException) {
+					// Ignore concurrent topic creation - we're good as long as topic was created and exists
```
Comment (elefeint) [inline spring-cloud-gcp-pubsub-stream-binder/src/main/java/org/springframework/cloud/gcp/stream/binder/pubsub/provisioning/PubSubChannelProvisioner.java:126]: Could you log a warning here?

**Comment by pivotal-issuemaster** [comment]
@mvmn Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-gcp&pullRequestId=2411)!

File: spring-cloud-gcp-pubsub-stream-binder/src/main/java/org/springframework/cloud/gcp/stream/binder/pubsub/provisioning/PubSubChannelProvisioner.java (hunk 1 above)
Comment (mvmn) [reply spring-cloud-gcp-pubsub-stream-binder/src/main/java/org/springframework/cloud/gcp/stream/binder/pubsub/provisioning/PubSubChannelProvisioner.java:126]: Sure

**Comment by mvmn** [comment]
Thank you @elefeint. 

I've added log warning and a unit test, please re-check if it's ok.

Thanks!

**Comment by mvmn** [comment]
Wait, apparently there are checkstyle errors now - let me fix that first...

File: spring-cloud-gcp-pubsub-stream-binder/src/main/java/org/springframework/cloud/gcp/stream/binder/pubsub/provisioning/PubSubChannelProvisioner.java (hunk 1 above)
Comment (meltsufin) [reply spring-cloud-gcp-pubsub-stream-binder/src/main/java/org/springframework/cloud/gcp/stream/binder/pubsub/provisioning/PubSubChannelProvisioner.java:126]: Why warning level if it's not a real issue?
I would go with DEBUG, or at most INFO.

**Comment by mvmn** [comment]
I can see the warning message in build log - which means unit test has worked and the code in catch clause was called.

Not sure why codecov is reporting a 7% drop in coverage here...

**Comment by elefeint** [comment]
Codecov is showing a decrease because the baseline includes integration tests; this is not a concern.

The warning label was because I asked for it; we can definitely downgrade to INFO. I was thinking it could be useful in production to notice it (and production typically does not show below WARN) and realize that a permanent, non-auto-created, topic would be a better solution where multiple load-balanced publishers exist.

File: spring-cloud-gcp-pubsub-stream-binder/src/main/java/org/springframework/cloud/gcp/stream/binder/pubsub/provisioning/PubSubChannelProvisioner.java (hunk 1 above)
Comment (mvmn) [reply spring-cloud-gcp-pubsub-stream-binder/src/main/java/org/springframework/cloud/gcp/stream/binder/pubsub/provisioning/PubSubChannelProvisioner.java:126]: Changed level to INFO according to @elefeint comment

File: spring-cloud-gcp-pubsub-stream-binder/src/test/java/org/springframework/cloud/gcp/stream/binder/pubsub/provisioning/PubSubChannelProvisionerTests.java [hunk 2]
```diff
@@ -176,4 +177,14 @@ public void testAfterUnbindConsumer_nonAnonymous() {
...
 	}
+
+	@Test
+	public void testProvisionConsumerDestination_concurrentTopicCreation() {
+		when(this.pubSubAdminMock.createTopic(any())).thenThrow(AlreadyExistsException.class);
+		when(this.pubSubAdminMock.getTopic(eq("non_existing_topic"))).thenReturn(null);
+
+		// Ensure no exceptions occur if topic already exists on create call
+		this.pubSubChannelProvisioner
+				.provisionConsumerDestination("non_existing_topic", "group1", this.properties);
```
Comment (elefeint) [inline spring-cloud-gcp-pubsub-stream-binder/src/test/java/org/springframework/cloud/gcp/stream/binder/pubsub/provisioning/PubSubChannelProvisionerTests.java:188]: Extreme nit: replace "non_existing_topic" with "already_existing_topic".

Comment (mvmn) [reply spring-cloud-gcp-pubsub-stream-binder/src/test/java/org/springframework/cloud/gcp/stream/binder/pubsub/provisioning/PubSubChannelProvisionerTests.java]: Done

**Comment by mvmn** [comment]
All points addressed, please re-review.

**Pull request merged**